# Effective length normalization for count based methods

The count of RNA-seq reads which can be uniquely assigned to a gene
depends on a number of factors, including the abundance of that gene,
the library size, the length of the gene,
sequence biases, the distribution of fragment lengths and others. Including a
model parameter to account for library size, then comparing counts across samples assumes
that the gene-specific factors other than gene abundance do not change
across samples: for example the length of the gene or the
gene-specific effects of sequence biases. These factors can
also be included as model parameters into a count based method,
through the use of a gene-by-sample matrix of normalization factors,
or an offset matrix (on the log scale).

Here we show how to use the estimates of effective length from the
output of RSEM and Cufflinks and use these as normalization for count
base methods DESeq2 and edgeR.

By gene length, or the
*effective length* of a gene, we refer to a gene's average transcript
length, averaging with respect to the abundance of each
transcript. For example, if a gene has two transcripts, one of length
1000 which accounts for 1/4 of the abundance, and one of length 2000
which accounts for 3/4 of the abundance, the effective length of
average transcript length would be $0.25(1000) + 0.75(2000) = 1750$

## Experiment files

The files are RNA-seq runs from the UCSD Human Reference Epigenome
Mapping Project. Note that these are individuals runs from experiments
which might have had multiple runs per experiment.

The files are in the order: left ventricle, sigmoid colon, left ventricle, sigmoid colon.

```{r}
(samples <- list.files("rsem_out"))
cond <- c("LV","SC","LV","SC")
```

```{r cache=TRUE}
rsem_list <- list()
for (i in 1:4) {
  cat(i)
  rsem_raw <- read.delim(paste0("rsem_out/",samples[i],"/",samples[i],".genes.results"))
  rsem_list[[i]] <- data.frame(gene_id=rsem_raw$gene_id, eff_length=rsem_raw$effective_length)
}
```

```{r}
head(rsem_list[[1]])
```

```{r cache=TRUE}
library(data.table)
library(dplyr)
cuff_list <- list()
for (i in 1:4) {
  cat(i)
  suppressWarnings({cuff_raw <- fread(paste0("cufflinks_out/",samples[i],"/isoforms.fpkm_tracking"))})
  cuff_raw <- as.data.frame(cuff_raw)
  cuff_list[[i]] <- data.frame(cuff_raw %>%
                               group_by(gene_id) %>%
                               summarize(eff_length=sum(length*FPKM)/sum(FPKM)))
}
```

```{r}
head(cuff_list[[1]])
```

```{r cache=TRUE}
rsem_summary <- data.frame(samp1=rsem_list[[1]]$eff_length, row.names=rsem_list[[1]]$gene_id)
for (i in 2:4) {
  stopifnot(all.equal(rownames(rsem_summary), as.character(rsem_list[[i]]$gene_id)))
  rsem_summary <- cbind(rsem_summary, newcol=rsem_list[[i]]$eff_length)
  names(rsem_summary)[i] <- paste0("samp",i)
}
```

```{r}
head(rsem_summary)
```

```{r cache=TRUE}
cuff_summary <- data.frame(samp1=cuff_list[[1]]$eff_length, row.names=cuff_list[[1]]$gene_id)
for (i in 2:4) {
  stopifnot(all.equal(rownames(cuff_summary), as.character(cuff_list[[i]]$gene_id)))
  cuff_summary <- cbind(cuff_summary, newcol=cuff_list[[i]]$eff_length)
  names(cuff_summary)[i] <- paste0("samp",i)
}
```

```{r}
head(cuff_summary)
```

```{r cache=TRUE}
idx <- intersect(rownames(rsem_summary), rownames(cuff_summary))
length(idx)
rsem_summary <- rsem_summary[idx,]
cuff_summary <- cuff_summary[idx,]
stopifnot(all.equal(rownames(rsem_summary), rownames(cuff_summary)))
colnames(rsem_summary) <- cond
colnames(cuff_summary) <- cond
```

```{r scatter}
library(rafalib)
mypar(2,2)
for (i in 1:4) {
  suppressWarnings(plot(rsem_summary[,i], cuff_summary[,i],log="xy",cex=.3,main=samples[i],
       xlab="rsem", ylab="cufflinks"))
  abline(0,1,col="blue")
}
```

```{r cache=TRUE, rsem_pairs}
#
suppressWarnings(pairs(rsem_summary,log="xy",cex=.1))
```

```{r cache=TRUE, cuff_pairs}
#
suppressWarnings(pairs(cuff_summary,log="xy",cex=.1))
```

```{r rsem_ma}
mypar()
rsem_a <- rowMeans(rsem_summary)
rsem_m <- log2(rowMeans(rsem_summary[,c(2,4)])/rowMeans(rsem_summary[,c(1,3)]))
plot(rsem_a, rsem_m, log="x", xlab="A", ylab="M", main="rsem", cex=.3)
abline(h=0,col="blue")
```

```{r cuff_ma}
mypar()
cuff_a <- rowMeans(cuff_summary)
cuff_m <- log2(rowMeans(cuff_summary[,c(2,4)])/rowMeans(cuff_summary[,c(1,3)]))
plot(cuff_a, cuff_m, log="x", xlab="A", ylab="M", main="cufflinks", cex=.3)
abline(h=0,col="blue")
```

```{r m_vs_m}
mypar()
plot(rsem_m, cuff_m, xlab="rsem", ylab="cufflinks", main="M vs M", cex=.3)
legend("bottomright",legend=paste0("pearson=",round(cor(rsem_m,cuff_m,use="complete"),2)),inset=.05)
abline(0,1,col="blue")
```

```{r m_minus_m}
mypar()
plot((rsem_a + cuff_a)/2, rsem_m - cuff_m, xlab="(rsem A + cufflinks A)/2", ylab="rsem M - cufflinks M",
     main="M minus M", cex=.3, log="x")
abline(h=0,col="blue")
```
